### ISSUE TITLE
Multicompiler with multiple output paths breaks WDS

### ISSUE THREAD

**Issue by SpaceK33z** [webpack/webpack-dev-server#641]
_Moved from https://github.com/webpack/webpack/issues/3091 - by @drphelps:_

**I'm submitting a feature request**

**Webpack version:**
2.x

**Please tell us about your environment:**
OSX 10.x

**Current behavior:**
Currently, a Multicompiler instance with a different `output.path` specified for each config will boil the `outputPath` property down to a "common" path. I'm referring to this block of code https://github.com/webpack/webpack/blob/547295ca0e91dd5bb2c446e2df19c887c6aa053c/lib/MultiCompiler.js#L51-L63. This breaks WDS filename resolution.

**Expected/desired behavior:**
WDS should serve resources from both paths or at least offer the ability to choose the first/preferred output.path when constructing the filename. I need to be able to output my files to separate (not necessarily nested) directories without breaking resolution.
- **What is the motivation / use case for changing the behavior?**
  In my setup, I am trying to build two separate apps in parallel and I need to be able to share resources between the two, including async modules. Consider the following config:

``` js
module.exports = [
  {
    context: __dirname,
    entry: "./app.js",
    output: {
      filename: "app.js",
      path: path.resolve(__dirname, './app/dist'),
      publicPath: '/static/javascript/dist/'
    },
    ...
  },
  {
    context: __dirname,
    entry: "./widget.js",
    output: {
      filename: "widget.js",
      path: path.resolve(__dirname, './widget/dist'),
      publicPath: '/widget/dist/'
    },
    ...
  }
]
```

**Comment by iMoses** [comment]
Just found that the hard way.

My use case is two configs, one for JavaScript and the other for static CSS, each requires a different output and resolves so I am using multi-compiler mode. I would like to be able to run webpack-dev-server simultaneously for both of them, using the same host & port configuration.

Is anybody working on it or should I take initiative? :)

**Comment by SpaceK33z** [comment]
> Is anybody working on it or should I take initiative? :)

Nope, feel free.

**Comment by iMoses** [comment]
@SpaceK33z 
I pushed a quick fix which does the trick for me. 
Let me know what you think.

**Comment by rainercedric23** [comment]
Any updates on @iMoses  fix? this is my issue also when I tried to run with multiple outputs. hopefully it can be pushed.

**Comment by SpaceK33z** [comment]
See #151

**Comment by ywmail** [comment]
This bug blocks me to use the WDS, but I found a solution from https://github.com/glenjamin/webpack-hot-middleware, the `webpack-hot-middleware` allows me to use multi compire instances.
```
const firstConfig = require('./config/first');
const secondConfig = require('./config/second');

let express = require('express');
let middleware = require('webpack-dev-middleware');
let app = express();

// Dev Server
[firstConfig, secondConfig].forEach(function (config) {
    let compiler = webpack(config);
    app.use(middleware(compiler, {
        publicPath: config.output.publicPath
    }));

    // Enables HMR
    app.use(webpackHotMiddleware(compiler, {
        log: console.log, path: config.output.publicPath + '__webpack_hmr', heartbeat: 10 * 1000
    }));

});


let server = app.listen(18088);

```

**Comment by rainercedric23** [comment]
Hi @ywmail, can I see your firstConfig and second config file for reference. many thanks.

**Comment by ywmail** [comment]
part of the code:
# server.js
```
const telcoConfig = require('./config/webpack.dev')({ env: 'development', app: 'telco' });
const marketplaceConfig = require('./config/webpack.dev')({ env: 'development', app: 'marketplace' });
```
# webpack.dev.js
```
    return webpackMerge(commonConfig({ env: ENV, app: app }), {

        entry: {

            'polyfills': ['./src/' + app + '/polyfills.browser.ts', hotMiddlewareScript],
            'vendor': ['./src/' + app + '/vendor.browser.ts', hotMiddlewareScript],
            'main': ['./src/' + app + '/main.browser.ts', hotMiddlewareScript]

        },

        devtool: 'cheap-module-source-map',

        output: {

            path: helpers.root('dist/' + app),

            publicPath: '/' + app + '/',

            filename: '[name].bundle.js',

            sourceMapFilename: '[name].map',

            chunkFilename: '[id].chunk.js',

            library: 'ac_[name]',
            libraryTarget: 'var',
        },

        plugins: [
        ],

    })
```

**Comment by akinnee** [comment]
Any update on this?

**Comment by elygre** [comment]
This would be very nice to have!

We are trying to fit webpack into an existing toolchain, requiring multiple configurations. The real killer would be to have a webpack-dev-server serving the assets from both the configs, at the paths specified by each.

**Comment by robgha01** [comment]
I just found out the hard way >.< any updates on this ? for now i guess i will have to not use this and stick to one embedded theme ...

**Comment by SpaceK33z** [comment]
Instead of constantly asking for updates, you can help. PR https://github.com/webpack/webpack-dev-middleware/pull/151 is a good starting point, but was never finished.

**Comment by NogsMPLS** [comment]
Looks like this might be fixed with this merge into dev-middleware: https://github.com/webpack/webpack-dev-middleware/pull/187

**Comment by maciej-gurban** [comment]
I'm using webpack@2.6.1 and webpack-dev-server@2.4.4 versions, and having two webpack configs - each with different context and devServer.contentBase - still doesn't seem to work. Is it supposed to be supported?

**Comment by maciej-gurban** [comment]
@faceyspacey Since you've worked on this, is there any docs I could read to understand how to properly configure webpack to use the multicompiler settings? I believe my use case is exactly as what @SpaceK33z described. Or is that workflow supported at all?

**Comment by jonohayon** [comment]
Hi, was there any progress on this?
I see that webpack/webpack-dev-middleware#187 fixed this problem in the middleware part, so that's not a problem anymore, and the code actually looks like it's supposed to be able to run multi-compiler configurations -- anyone knows what may cause this to not work (as it looks like it should)?

**Comment by shellscape** [comment]
#187 resolved the root issue that @SpaceK33z moved over from webpack. If there are errors being thrown please do report it as a bug separately and we can triage. If there's only difficulty in getting it to work, please head to StackOverflow or the webpack gitter for more discussion and peer troubleshooting.

**Comment by jonaskello** [comment]
If you want to have the *same* `output.publicPath` for both your configs, or have them nested so one is the root and the other is a sub, then you need to make sure the configs have the exact same setting for `output.path` otherwise only the first config will have effect. Not sure why this is, I spent several hours trial and error to find this out so I just wanted to mention it for future reference.

**Comment by SergeShkurko** [comment]
My solution to a similar problem with `webpack multi compiller` & `react` & `HMR`

I have 2 sub projects: app & widget

App located at `localhost:8080/*` (except `localhost:8080/widget`)
Widget located at `localhost:8080/widget`

### webpack.config.js
```javascript
module.exports = [{
  entry: {
    app: [
      'src/app/index.js',
      // path = output.publicPath + __webpack_hmr
      'webpack-hot-middleware/client?path=/__webpack_hmr&timeout=20000&reload=true',
    ],
  },
  output: {
    publicPath: '/',
  },
  plugins: [
    new HtmlWebpackPlugin({
      template: './src/app/index.html',
    }),
    new webpack.HotModuleReplacementPlugin(),
  ],
  // Other module 1 options
},{
  entry: {
    widget: [
      'src/app/index.js',
      // path = output.publicPath + __webpack_hmr !see the difference between entries!
      'webpack-hot-middleware/client?path=/widget__webpack_hmr&timeout=20000&reload=true',
    ],
  },
  output: {
    publicPath: '/widget',
  },
  plugins: [
    new HtmlWebpackPlugin({
      template: './src/widget/index.html',
    }),
    new webpack.HotModuleReplacementPlugin(),
  ],
  // Other module 2 options
}]
```

### dev-server.js
```javascript
const
  fs = require('fs'),
  http = require('http'),
  express = require('express'),
  webpack = require('webpack')

const app = express()

;(function() {
  const webpackConfig = require(process.env.WEBPACK_CONFIG || './webpack.config')

  webpackConfig.forEach((config, index) => {
    const compiler = webpack(config)

    app
      .use(require('webpack-dev-middleware')(compiler, {
        publicPath: config.output.publicPath,
      }))
      .use(require('webpack-hot-middleware')(compiler, {
        log: console.log,
        path: `${config.output.publicPath}__webpack_hmr`,
        heartbeat: 10 * 1000,
      }))
  })
})()

if (require.main === module) {
  const server = http.createServer(app)
  server.listen(process.env.PORT || 8080, () =>
    console.log('Listening on %j', server.address())
  )
}
```

And... add `react-hot-loader` in app & widget entrypoints

Works great 👍

**Comment by hytromo** [comment]
@jonaskello I had the same problem. It was created by me sharing data between the different configs returned.

My guess would be that webpack is not acting with an immutable way on this configs once you return them, so if you are sharing data between the config, when the first config is being parsed and mutated, part of the 2nd one is, too, and that in turn makes it invalid etc.

By not sharing data between the configs, it seems that it is working for any output or public path I set.
